File: mvpboilerplate/src/main/java/com/agit/lutluthfi/mvpboilerplate/base/PlateBaseActivity.java
package com.agit.lutluthfi.mvpboilerplate.base;

import android.app.ProgressDialog;
import android.content.Context;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;
import android.widget.Toast;

import com.agit.lutluthfi.mvpboilerplate.R;
import com.agit.lutluthfi.mvpboilerplate.utils.CommonUtils;
import com.agit.lutluthfi.mvpboilerplate.utils.KeyboardUtils;
import com.agit.lutluthfi.mvpboilerplate.utils.NetworkUtils;

public abstract class PlateBaseActivity extends AppCompatActivity implements IPlateBaseView {

    private Snackbar mSnackbar;
    private ProgressDialog mProgressDialog;

    protected abstract void setupView();

    protected abstract void setupListener();

    @Override
    public void showLoading() {
        hideLoading();
        mProgressDialog = CommonUtils.showLoadingDialog(this);
    }

    @Override
    public void hideLoading() {
        if (mProgressDialog != null && mProgressDialog.isShowing()) mProgressDialog.cancel();
    }

    @Override
    public boolean isLoading() {
        return mProgressDialog != null && mProgressDialog.isShowing();
    }

    @Override
    public boolean isNetworkConnected() {
        return NetworkUtils.isNetworkConnected(this);
    }

    @Override
    public void onError(String message) {
        if (message != null && !message.isEmpty()) {
            showSnackbar(message, Snackbar.LENGTH_SHORT, R.color.colorWhite);
        } else {
            showSnackbar(getString(R.string.utils_error_view), Snackbar.LENGTH_SHORT, R.color.colorWhite);
        }
    }

    @Override
    public void onError(int resId) {
        showSnackbar(getString(resId), Snackbar.LENGTH_SHORT, R.color.colorWhite);
    }

    @Override
    public void showMessage(String message) {
        if (message != null && !message.isEmpty()) {
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, getString(R.string.utils_error_view), Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void showMessage(int resId) {
        showMessage(getString(resId));
    }

    @Override
    public void printLog(String tag, String message) {
        if (message != null && !message.isEmpty()) Log.d(tag, message);
        else Log.d(tag, getString(R.string.utils_error_view));
    }

    @Override
    public void printLog(String tag, int resId) {
        printLog(tag, getString(resId));
    }

    @Override
    public void printLog(String tag, Throwable tr) {
        Log.d(tag, tr.getMessage(), tr);
    }

    @Override
    public void showSnackbar(String message, int duration, int textColor) {
        hideSnackbar();
        createSnackbar(message, duration, textColor).show();
    }

    @Override
    public void showSnackbar(int resId, int duration, int textColor) {
        hideSnackbar();
        createSnackbar(getString(resId), duration, textColor).show();
    }
    @Override
    public void showSnackbarWithAction(String message, int duration, int textColor, String action, View.OnClickListener listener) {
        hideSnackbar();
        createSnackbar(message, duration, textColor).setAction(action, listener).show();
    }

    @Override
    public void showSnackbarWithAction(int resId, int duration, int textColor, String action, View.OnClickListener listener) {
        hideSnackbar();
        createSnackbar(getString(resId), duration, textColor).setAction(action, listener).show();
    }

    @Override
    public void hideSnackbar() {
        if (mSnackbar != null && mSnackbar.isShown()) mSnackbar.dismiss();
    }

    @Override
    public boolean isSnackbarShowing() {
        return mSnackbar != null && mSnackbar.isShown();
    }

    private Snackbar createSnackbar(String message, int duration, int textColor) {
        if (message != null && !message.isEmpty()) {
            mSnackbar = Snackbar.make(findViewById(android.R.id.content), message, duration);
            View snackbarView = mSnackbar.getView();
            TextView textView = snackbarView.findViewById(android.support.design.R.id.snackbar_text);
            textView.setTextColor(ContextCompat.getColor(this, textColor));
        } else {
            mSnackbar = Snackbar.make(findViewById(android.R.id.content), getString(R.string.utils_error_common), duration);
            View snackbarView = mSnackbar.getView();
            TextView textView = snackbarView.findViewById(android.support.design.R.id.snackbar_text);
            textView.setTextColor(ContextCompat.getColor(this, textColor));
        }
        return mSnackbar;
    }
}

File: mvpboilerplate/src/main/java/com/agit/lutluthfi/mvpboilerplate/base/PlateBaseActivity.kt
package com.agit.lutluthfi.mvpboilerplate.base

import android.app.ProgressDialog
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.TextView
import android.widget.Toast

import com.agit.lutluthfi.mvpboilerplate.R
import com.agit.lutluthfi.mvpboilerplate.util.CommonUtils
import com.agit.lutluthfi.mvpboilerplate.util.NetworkUtils

abstract class PlateBaseActivity : AppCompatActivity(), IPlateBaseView {

    private var snackbar: Snackbar? = null
    private var progressDialog: ProgressDialog? = null

    protected abstract fun setupView()

    protected abstract fun setupListener()

    override fun showLoading() {
        runOnUiThread {
            if (isLoading()) hideLoading()
            progressDialog = CommonUtils.showLoadingDialog(context = this)
        }
    }

    override fun hideLoading() {
        runOnUiThread { progressDialog?.run { if (isShowing) cancel() } }
    }

    override fun isLoading(): Boolean {
        var isShowing = false
        runOnUiThread { progressDialog?.run { isShowing = this.isShowing } }
        return isShowing
    }

    override fun isNetworkConnected(): Boolean = NetworkUtils.isNetworkConnected(context = this)

    override fun showMessage(message: String?) {
        val viewError = getString(R.string.utils_error_view)
        message?.let {
            runOnUiThread {
                if (it.isNotEmpty()) {
                    Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this, viewError, Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    override fun showMessage(resId: Int) = showMessage(getString(resId))

    override fun printLog(tag: String, message: String?) {
        message?.let {
            return@let if (it.isNotEmpty()) Log.d(tag, it)
            else Log.d(tag, getString(R.string.utils_error_view))
        }
    }

    override fun printLog(tag: String, resId: Int) = printLog(tag, getString(resId))

    override fun printLog(tag: String, tr: Throwable?) {
        tr?.let { Log.d(tag, it.message, it) }
    }

    override fun showSnackbar(message: String?, duration: Int, textColor: Int) {
        message?.let {
            runOnUiThread {
                hideSnackbar()
                createSnackbar(it, duration, textColor)?.run { show() }
            }
        }
    }

    override fun showSnackbar(resId: Int, duration: Int, textColor: Int) {
        runOnUiThread {
            if (isSnackbarShowing()) hideSnackbar()
            createSnackbar(getString(resId), duration, textColor)?.run { show() }
        }
    }

    override fun showSnackbarWithAction(message: String?,
                                        duration: Int,
                                        textColor: Int,
                                        action: String,
                                        listener: View.OnClickListener) {
        message?.let {
            runOnUiThread {
                if (isSnackbarShowing()) hideSnackbar()
                createSnackbar(it, duration, textColor)?.run {
                    setAction(action, listener)
                    show()
                }
            }
        }
    }

    override fun showSnackbarWithAction(resId: Int,
                                        duration: Int,
                                        textColor: Int,
                                        action: String,
                                        listener: View.OnClickListener) {
        runOnUiThread {
            if (isSnackbarShowing()) hideSnackbar()
            createSnackbar(getString(resId), duration, textColor)?.run {
                setAction(action, listener)
                show()
            }
        }
    }

    override fun hideSnackbar() {
        runOnUiThread { snackbar?.run { if (isShown) dismiss() } }
    }

    override fun isSnackbarShowing(): Boolean {
        var isShowing = false
        snackbar?.run { isShowing = this.isShown }
        return isShowing
    }

    private fun createSnackbar(message: String?, duration: Int, textColor: Int): Snackbar? {
        val color = ContextCompat.getColor(this@PlateBaseActivity, textColor)
        val commonError = getString(R.string.utils_error_common)
        message?.let {
            if (it.isNotEmpty()) {
                snackbar = Snackbar.make(findViewById(android.R.id.content), it, duration)
                val snackbarView = snackbar?.run { view }
                return@let snackbarView?.run {
                    val textView = findViewById<TextView>(android.support.design.R.id.snackbar_text)
                    textView.setTextColor(color)
                }
            } else {
                snackbar = Snackbar.make(findViewById(android.R.id.content), commonError, duration)
                val snackbarView = snackbar?.run { view }
                snackbarView?.run {
                    val textView = findViewById<TextView>(android.support.design.R.id.snackbar_text)
                    return@let textView.setTextColor(color)
                }
            }
        }
        return snackbar as Snackbar
    }
}
